GSoC 2025 blog on the ABI lowering library (#76)
llvm/llvm-blog-www: content/posts/2025-08-25-abi-library.md

---
author: "Narayan Sreekumar (vortex73)"
date: "2025-08-25"
tags: ["GSoC", "abi", "codegen", "sysv"]
title: "GSoC 2025: Introducing an ABI Lowering Library"
---

# Introduction

> In this post I'm going to outline details about a new ABI lowering library I've been developing for LLVM as part of GSoC 2025! The aim was to extract the ABI logic from clang and create a reusable library that any LLVM frontend can use for correct C interoperability.

Reviewer: ```suggestion
In this post I'm going to outline details about a new ABI lowering library I've been developing for LLVM as part of GSoC 2025! The aim was to extract the ABI logic from Clang and create a reusable library that any LLVM frontend can use for correct C interoperability.
```

</div>

## Independent ABI Type System

At the heart of the library is `llvm::abi::Type`, a type system designed specifically for ABI decisions:

```cpp
class Type {
protected:
  TypeKind Kind;
  TypeSize SizeInBits;
  Align ABIAlignment;

public:
  TypeKind getKind() const { return Kind; }
  TypeSize getSizeInBits() const { return SizeInBits; }
  Align getAlignment() const { return ABIAlignment; }

  bool isInteger() const { return Kind == TypeKind::Integer; }
  bool isStruct() const { return Kind == TypeKind::Struct; }
  // ... other predicates that matter for ABI
};
```

It contains **more information than LLVM IR types** (which for instance doesn't distinguish between `__int128` and `_BitInt(128)`, both just `i128`), but **less information than frontend types** like Clang's QualType (which carry parsing context, sugar, and other frontend-specific concerns that don't matter for calling conventions).
```cpp
class IntegerType : public Type {
private:
  bool IsSigned;
  bool IsBitInt;        // Crucially different from __int128!

public:
  IntegerType(uint64_t BitWidth, Align Align, bool Signed,
              bool BitInt = false);
};
```

## Frontend-to-ABI Mapping

> The QualTypeMapper class handles the complex job of converting frontend types to ABI types. Here's how it tackles C++ inheritance:

Reviewer: Something I would like to highlight here is that the type mapping is expected to be simple for C-only users, but complex for Clang in particular because it has to support C++. Something like this (I pulled the paragraph at the end of the section to the front):

```suggestion
The QualTypeMapper class handles the job of converting Clang frontend types to ABI types.

The ABI library is primarily intended to handle the C ABI. The C type system is relatively simple, and as such the type mapping from frontend types to ABI types is simple as well.

However, Clang also needs support for the C++ ABI. The type mapping for this case is significantly more complicated, as C++ object layout involves vtables, base class subobjects, virtual inheritance, and all sorts of edge cases that need to be preserved for correct ABI decisions.

Here is an excerpt from the C++ inheritance handling:
```

---
author: "Narayan Sreekumar (vortex73)"

> date: "2025-10-27"

Reviewer 2: ```suggestion
date: "2025-11-03"
```
